File: src/main/scala/gs2vs/core/db/DataSource.scala
package gs2vs.core.db

import com.zaxxer.hikari.{HikariConfig, HikariDataSource}
import gs2vs.core.JdbcConfig
import org.slf4j.{Logger, LoggerFactory}

import java.sql.Connection

trait DataSource {
  def getConnection(): Connection
}

class DataSourceImpl(jdbcConfig: JdbcConfig) extends DataSource {
  protected val logger: Logger = LoggerFactory.getLogger(classOf[DataSource])

  private val hikariConfig = new HikariConfig()
  hikariConfig.setJdbcUrl(jdbcConfig.url)
  hikariConfig.setUsername(jdbcConfig.username)
  jdbcConfig.password.foreach(pwd => hikariConfig.setPassword(pwd))
  hikariConfig.setConnectionTimeout(3000)
  hikariConfig.setConnectionInitSql("select 1")
  hikariConfig.setInitializationFailTimeout(3000)
  hikariConfig.addDataSourceProperty("cachePrepStmts", "true")
  hikariConfig.addDataSourceProperty("prepStmtCacheSize", "250")
  hikariConfig.addDataSourceProperty("prepStmtCacheSqlLimit", "2048")

  private lazy val ds =
    try { new HikariDataSource(hikariConfig) }
    catch {
      case e: Throwable =>
//        logger.error(s"init datasource fail. jdbcConfig: ${jdbcConfig}", e)
        throw e
    }

  def getConnection(): Connection = {
    ds.getConnection()
  }

}

File: src/main/java/gs2vs/core/repository/DataSource.java
//package gs2vs.core.repository;
//
//import com.zaxxer.hikari.HikariConfig;
//import com.zaxxer.hikari.HikariDataSource;
//
//import java.sql.Connection;
//import java.sql.SQLException;
//
//public class DataSource {
//
//    private static HikariConfig config = new HikariConfig();
//    private static HikariDataSource ds;
//
//    static {
//        config.setJdbcUrl( "jdbc:postgresql://localhost/d2emo" );
//        config.setUsername( "lorance" );
////        config.setPassword( "" );
//        config.setConnectionTimeout(3000);
//        config.setConnectionInitSql("select 1");
//        config.setInitializationFailTimeout( 3000 );
//        config.addDataSourceProperty( "cachePrepStmts" , "true" );
//        config.addDataSourceProperty( "prepStmtCacheSize" , "250" );
//        config.addDataSourceProperty( "prepStmtCacheSqlLimit" , "2048" );
//        ds = new HikariDataSource( config );
//    }
//
//    private DataSource() {}
//
//    public static Connection getConnection() throws SQLException {
//        return ds.getConnection();
//    }
//}
